### PR TITLE
fix(azure): add support for Azure affinity groups

### PR DESCRIPTION
**Pull request by gvilarino**
This PR adds two fixes that will make a 5-vm Deis cluster stable on Windows Azure:

#### Enhancements
1. Support for specifying `affinity-group` as a valid argument to the `azure-coreos-cluster` script.
  * Overall platform startup speed reduced by about 15 minutes (ballpark) for a 5-vm cluster on Azure.
  * Overall platorm stability and inter-fleet latency reduced to an average of 59ms. [See full cluster stats](https://gist.github.com/gvilarino/aa20c427b5e08b959083)
2. Specific settings for `etcd.service` that will prevent the cluster from dying because of reaching the file-open limit
  * Cluster becomes stable and stops throwing "too many open files" error anymore (possibly related to #1240).

#### TODO
* Add setting an `affinity-group` option as a **very recommended** option in the official Deis docs for an Azure setup.

#### Acknowlegements
Enormous thanks to @aledbf , @sedouard and @carmstrong for helping me with the cluster setup and debugging my situation.

### CODE REVIEW

**Comment by carmstrong** [comment]
> Add setting an affinity-group option as a very recommended option in the official Deis docs for an Azure setup.

Let's definitely do that :thumbsup:.

I'm hesitant to overwrite the etcd config, because then we'll have closely watch the default etcd settings to ensure we're not overriding any of them. Here's the default etcd.service:
```
[Unit]
Description=etcd

[Service]
User=etcd
PermissionsStartOnly=true
Environment=ETCD_DATA_DIR=/var/lib/etcd
Environment=ETCD_NAME=%m
ExecStart=/usr/bin/etcd
Restart=always
RestartSec=10s
```

I'm wondering if we should instead PR the file handles change upstream to CoreOS directly.

**Comment by aledbf** [comment]
@carmstrong the change in the service
```
        LimitNOFILE=1048576
        LimitNPROC=1048576
```

is related to the first cluster that @gvilarino had. In that cluster (5 nodes) the error in etcd caused the election a new leader every a couple of seconds and this opened a lot of connections causing the error "too many open files" in etcd and because of this fleet was unusable.

The particular error was heartbeat timeouts that until now are solved using an affinity-group.

**Comment by aledbf** [comment]
@carmstrong and Azure adds a custom oem service related to etcd
```
/run/systemd/system/etcd.service.d/10-oem.conf
  [Service]
  Environment=ETCD_PEER_ELECTION_TIMEOUT=1200
```

**Comment by carmstrong** [comment]
Thanks @aledbf - I understand the change, and think it's a good one! I'd just rather have it upstream in CoreOS so it's not something we have to manage ourselves.

**Comment by gvilarino** [comment]
Maybe for the time being (until it's added to CoreOS) it's a viable fix? Deis won't be stable on Azure without any of these upgrades =/

**Comment by carmstrong** [comment]
@marineam Said we could PR the changes to the etcd service upstream at https://github.com/coreos/coreos-overlay/blob/master/dev-db/etcd/files/etcd.service

> Maybe for the time being (until it's added to CoreOS) it's a viable fix?
I'm fine with that once this is merged upstream -- then we know it's going to match what's upstream, even if we're not there yet.

**Comment by gvilarino** [comment]
@carmstrong would you like me to revert f4c2049b2233a08cd05388cc8465149d60bf66d8 so you can PR the `affinity-group` feature? I can create a separate PR for `etcd.service` udates.

**Comment by carmstrong** [comment]
@gvilarino I think these are fine as-is, except you'll need to amend them to conform to our commit style guidelines: http://docs.deis.io/en/latest/contributing/standards/#commit-style

We can merge this once you add an additional commit adding some Azure documentation around using affinity groups, and that it's strongly recommended. I'll PR the upstream change to etcd.

**Comment by carmstrong** [comment]
> I'll PR the upstream change to etcd.

coreos/coreos-overlay#1124

**Comment by gvilarino** [comment]
@carmstrong do those look good? I'm sorry about the merge commit; I wasn't sure how to exclude it :*

**Comment by gvilarino** [comment]
I'll add the docs now

**Comment by gvilarino** [comment]
I'm done @carmstrong, please let me know if anything else is missing.

File: contrib/azure/azure-user-data-template [hunk 1]
```diff
@@ -25,3 +25,18 @@ coreos:
...
+        Description=etcd
+        [Service]
+        User=etcd
+        PermissionsStartOnly=true
+        Environment=ETCD_DATA_DIR=/var/lib/etcd
+        Environment=ETCD_NAME=%m
+        ExecStart=/usr/bin/etcd
+        Restart=always
+        RestartSec=10s
+        LimitNOFILE=1048576
```
Comment (carmstrong) [inline contrib/azure/azure-user-data-template]: Let's change `LimitNOFILE` to `LimitNOFILE=40000` and remove `LimitNPROC` so this matches the proposed upstream change in coreos/coreos-overlay#1124.

File: docs/installing_deis/azure.rst [hunk 2]
```diff
@@ -69,7 +69,7 @@ With the management certificate and cloud config in place, we are ready to creat
 
 * Create a container called ``vhds`` within a storage account in the same region as your cluster using the Azure portal. Note the URL of the container for the cluster creation script below.
 * Choose a cloud service name for your Deis cluster for the script below. The script will automatically create this cloud service for you.
-* Choose an Azure `region`_ to use. Supply it in quotes with the ``--location`` parameter. The default is "West US".
+* Create an `affinity group`_ if you already don't have one. Supply it in quotes with the ``--affinity-group`` parameter. Although *using an affinity group is not mandatory*, it is **highly recommended** since it tells the Azure fabric to place all VMs in the cluster physically close to each other, reducing inter-node latency by a great deal. If you don't want ot use affinity groups, specify a `region`_ for Azure to use with a ``--location`` parameter. The default is ``"West US"``. If you specify both parameters, ``location`` will be ignored. Please note that the script *will not* create an affinity group by itself; it expects the affinity group exists.
```
Comment (carmstrong) [inline docs/installing_deis/azure.rst]: Typo "ot" instead of "to", but otherwise this reads great.

**Comment by gvilarino** [comment]
@carmstrong all done!

**Comment by carmstrong** [comment]
Thanks @gvilarino! I'll manually verify this and then we should be good to :ship:.

**Comment by gvilarino** [comment]
Awesomeness!

:shipit:

**Comment by mboersma** [comment]
I don't currently have a way to test this, but code LGTM.

**Comment by gvilarino** [comment]
I tried this a lot on Azure. Actually @aledbf is quite aware of my tryings and he can vouch for it :D

**Comment by gvilarino** [comment]
Actually I just discovered a slight typo in my doc. Fixing it now.

**Comment by gvilarino** [comment]
Now it's fine :shipit:  :)

**Comment by carmstrong** [comment]
We just need to get Jenkins back online (in the next day or two) and let it run a full pass to confirm the etcd changes. I'm sure it's fine, but we're trying to be good about testing :)

**Comment by gvilarino** [comment]
Totally respect that. Our project is poor about testing and I'm bitchin' all the time :+1:

**Comment by carmstrong** [comment]
> Totally respect that. Our project is poor about testing and I'm bitchin' all the time

:heart:

**Comment by carmstrong** [comment]
@gvilarino Let's rebase this on master, and remove the etcd changes as they've been merged into our main cloud-config as of #3311.

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

**Comment by gvilarino** [comment]
@carmstrong all done!

**Comment by carmstrong** [comment]
@gvilarino Finally, can we squash the commits in one?

**Comment by carmstrong** [comment]
@gvilarino I think we can merge this shortly after the fix up. Thanks for doing this!

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

**Comment by gvilarino** [comment]
Crap, this squash/rebase is proving impossible. I'll just start another one.

**Comment by gvilarino** [comment]
@carmstrong all done on https://github.com/deis/deis/pull/3329

Sorry but doing that was easier than rebasing after several merges.

**Pull request closed without merging**